### ISSUE TITLE
Suggestion of improvement for the section on records in the manual

### ISSUE THREAD

**Issue by esope** [ocaml/ocaml#12582]
The section in the manual that describes the constructs related to records says:

> The expression `{ expr with field1 [= expr1] ; … ; fieldn [= exprn] }` builds a fresh record with fields `field1 … fieldn` equal to `expr1 … exprn`, and all other fields having the same value as in the record `expr`. In other terms, it returns a shallow copy of the record `expr`, except for the fields `field1 … fieldn`, which are initialized to `expr1 ... exprn`.

While this is perfectly clear, I suggest the explanation could be slightly expanded by giving an example that involves mutable fields. Here is a tentative example:

```ocaml
type r = { mutable a: int; mutable b: bool }
let r1 = { a = 0; b = false }
let r2 = { r1 with b = false } (* this creates a fresh shallow copy of r1 *)
let () =
  r1.a <- 1;
  r1.b <- true
let () = (* the fields of r2 have not changed *)
  assert (r2.a = 0);
  assert (r2.b = false)
```

I think this example might surprise some programmers, that are not aware that `{ r1 with b = false }` allocates a new record.
If instead a programmer (mistakenly) thinks that each _mutable_ fields contains a pointer to a block, then she might expect that `r2.a` contains `1` in the example from above.
This example could help clarify the "builds a fresh record" part of the manual.

Alternatively, do you think a warning could be of help to programmers in such cases? (i.e., `{ expr with ... }` where `expr` contains mutable fields that are not redefined in the `...` expression)

**Comment by gasche** [comment]
I am not sure about a warning:
- presumably all *existing* code using the feature is written by people who are aware of the effect (otherwise you get into unexpected behaviors quite quickly)
- it is not clear that we want to discourage this pattern; do we?
- if we made it a warning and people wanted to avoid it, what code would we recommend writing instead, and is it reasonable to require this change?

(Regarding the extra examples in the manual: why not, but do programmers that get caught by this subtlety of mutable state actually read this section of the manual, before or after hitting the issue and being perplexed by it?)

Currently our warnings are mostly designed for the use-case "everyone or at least enough people consider that X is terrible style, activate me to prevent it in your project". We don't have warnings in the category "Expert programmers write this just fine and would not enable me, but if you are a beginner you may want a tip that you are doing something a bit advanced here!". I am not sure how we would go about that.

**Comment by esope** [comment]
Agreed about the warning. A better place for such a warning could be in a linter, that could be useful for people who are learning programming in OCaml.

About the extra example: it seems that the manual is not the best place to have it. I am fine with that.
What other place do you suggest?

**Comment by gasche** [comment]
Here would be my proposal:

```
In particular, the shallow copy of @'x'@ in the definition @'let' 'y'
'=' '{' 'x' 'with' \ldots '}'@ ensures that updates to a mutable field
(@'x.v' '<-' e@) do not affect the corresponding field @'y.v'@ of
@'y'@, and conversely. On the other hand, if a field of @'x'@ is
a reference (@"'a" 'ref'@), then @'y'@ will have the same reference
and any change (@'x.v' ':=' e@) will also be visible in @'y.v'@.
```

(cc @Octachron, who could help make decisions about manual updates)

**Comment by Octachron** [comment]
The first setence looks useful from my point of view. The second sentence feels a bit unrelated to the issue at hand, since this more the classical problem of nesting a mutable data inside an immutable container.
